controllers/openstackfloatingippool_controller.go

```diff
@@ -0,0 +1,410 @@
+					},
+					Address: ip,
+					Prefix:  32,
+				},
+			}
+
+			// If the ReclaimPolicy is Delete, we add the finalizer to the IPAddress if they are not pre-allocated
+			// this is to ensure that the IPAddress is deleted from openstack when the claim is deleted
+			if pool.Spec.ReclaimPolicy == infrav1.ReclaimDelete && !contains(pool.Spec.PreAllocatedFloatingIPs, ip) {
+				controllerutil.AddFinalizer(ipAddress, infrav1.DeleteFloatingIPFinalizer)
+			}
+
+			if err = r.Client.Create(ctx, ipAddress); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			claim.Status.AddressRef.Name = ipAddress.Name
+			if err = r.Client.Status().Update(ctx, &claim); err != nil {
+				log.Error(err, "Failed to update IPAddressClaim status", "claim", claim.Name, "ip", ip)
+				return ctrl.Result{}, err
+			}
+			scope.Logger().Info("Claimed IP", "ip", ip)
+		}
+	}
+	if err = r.setIPStatuses(ctx, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *OpenStackFloatingIPPoolReconciler) reconcileDelete(ctx context.Context, scope scope.Scope, pool *infrav1.OpenStackFloatingIPPool) error {
+	log := ctrl.LoggerFrom(ctx)
+	ipAddresses := &ipamv1.IPAddressList{}
+	if err := r.Client.List(ctx, ipAddresses, client.InNamespace(pool.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return err
+	}
+	// If there are still IPAddress objects that are not deleted, there are still claims on this pool and we should not delete it
+	// beause the pool is needed to clean up the addresses from openstack
+	if len(ipAddresses.Items) > 0 {
+		log.Info("Waiting for IPAddress to be deleted before deleting OpenStackFloatingIPPool")
+		return errors.New("waiting for IPAddress to be deleted, until we can delete the OpenStackFloatingIPPool")
+	}
+
+	networkingService, err := networking.NewService(scope)
+	if err != nil {
+		return err
+	}
+
+	// Clean up ips created by the pool
+	for _, ip := range diff(pool.Status.IPs, pool.Spec.PreAllocatedFloatingIPs) {
+		if err := networkingService.DeleteFloatingIP(pool, ip); err != nil {
+			return fmt.Errorf("delete floating IP: %w", err)
+		}
+	}
+
+	if controllerutil.RemoveFinalizer(pool, infrav1.OpenStackFloatingIPPoolFinalizer) {
+		log.Info("Removing finalizer from OpenStackFloatingIPPool")
+		return r.Client.Update(ctx, pool)
+	}
+	return nil
+}
+
+func union(a []string, b []string) []string {
+	m := make(map[string]struct{})
+	for _, item := range a {
+		m[item] = struct{}{}
+	}
+	for _, item := range b {
+		m[item] = struct{}{}
+	}
+	result := make([]string, 0, len(m))
+	for item := range m {
+		result = append(result, item)
+	}
+	return result
+}
+
+func diff(a []string, b []string) []string {
+	m := make(map[string]struct{})
+	for _, item := range a {
+		m[item] = struct{}{}
+	}
+	for _, item := range b {
+		delete(m, item)
+	}
+	result := make([]string, 0, len(m))
+	for item := range m {
+		result = append(result, item)
+	}
+	return result
+}
```

Comment:
I wonder if there's a potential to use this: https://github.com/kubernetes/utils/blob/master/set/set.go

Reply:
We could, but we would need to bump k8s.io/utils and  k8s.io/utils/pointer is deprecated in favor of k8s.io/utils/ptr so that would involve some refactoring and should probably be done in a separate PR

Reply:
Agreed!